fe/src/main/java/org/apache/doris/load/loadv2/dpp/MinimumCoverageRollupTreeBuilder.java

```diff
@@ -0,0 +1,126 @@
+// builded from the front index(eg: index3 is the child of index2). This can be
+// further optimized based on the row number of the index.
+public class MinimumCoverageRollupTreeBuilder implements RollupTreeBuilder {
+    public RollupTreeNode build(EtlJobConfig.EtlTable tableMeta) {
+        List<EtlJobConfig.EtlIndex> indexes = tableMeta.indexes;
+        List<EtlJobConfig.EtlIndex> indexMetas = new ArrayList<>();
+        EtlJobConfig.EtlIndex baseIndex = null;
+        for (EtlJobConfig.EtlIndex indexMeta : indexes) {
+            if (indexMeta.isBaseIndex) {
+                baseIndex = indexMeta;
+                continue;
+            }
+            indexMetas.add(indexMeta);
+        }
+        List<EtlJobConfig.EtlColumn> baseIndexColumns = baseIndex.columns;
+        List<String> baseKeyColumns = new ArrayList<>();
+        List<String> baseValueColumns = new ArrayList<>();
+        for (EtlJobConfig.EtlColumn columnMeta : baseIndexColumns) {
+            if (columnMeta.isKey) {
+                baseKeyColumns.add(columnMeta.columnName);
+            } else {
+                baseValueColumns.add(columnMeta.columnName);
+            }
+        }
+        RollupTreeNode root = new RollupTreeNode();
+        root.parent = null;
+        root.keyColumnNames = baseKeyColumns;
+        root.valueColumnNames = baseValueColumns;
+        root.indexId = baseIndex.indexId;
+        root.indexMeta = baseIndex;
+
+        // sort the index metas to make sure the column number decrease
+        Collections.sort(indexMetas, new EtlJobConfig.EtlIndexComparator().reversed());
+        boolean[] flags = new boolean[indexMetas.size()];
+        for (int i = 0; i < indexMetas.size(); ++i) {
+            flags[i] = false;
+        }
+        for (int i = 0; i < indexMetas.size(); ++i) {
+            List<String> keyColumns = new ArrayList<>();
+            List<String> valueColumns = new ArrayList<>();
+            for (EtlJobConfig.EtlColumn column : indexMetas.get(i).columns) {
+                if (column.isKey) {
+                    keyColumns.add(column.columnName);
+                } else {
+                    valueColumns.add(column.columnName);
+                }
+            }
+            insertIndex(root, indexMetas.get(i), keyColumns, valueColumns, i, flags);
+        }
+        return root;
+    }
+
+    // DFS traverse to build the rollup tree
+    private void insertIndex(RollupTreeNode root, EtlJobConfig.EtlIndex indexMeta,
+                             List<String> keyColumns,
+                             List<String> valueColumns, int id, boolean[] flags) {
+        if (root.children != null) {
+            for (RollupTreeNode child : root.children) {
+                insertIndex(child, indexMeta, keyColumns, valueColumns, id, flags);
+                if (flags[id]) {
+                    return;
+                }
+            }
+        }
+        if (flags[id]) {
```

Comment:
seems to be not needed

Reply:
Recursive happends when root node has children, so it is needed